File: ios/MapPolyline.swift
import GoogleMaps

class MapPolylineOptions {
  func buildPolyline(_ p: RNPolyline) -> GMSPolyline {
    let path = GMSMutablePath()
    p.coordinates.forEach {
      path.add(
        CLLocationCoordinate2D(latitude: $0.latitude, longitude: $0.longitude)
      )
    }
    let pl = GMSPolyline(path: path)
    if let w = p.width { pl.strokeWidth = CGFloat(w) }
    if let c = p.color?.toUIColor() { pl.strokeColor = c }
    pl.zIndex = Int32(p.zIndex)
    if let cap = p.lineCap {
      /// pl.lineCap = mapLineCap(cap)
    }
    if let join = p.lineJoin {
      /// pl.strokeJoin = mapLineJoin(join)
    }
    return pl
  }
}

func mapLineCap(_ t: RNLineCapType?) -> CGLineCap {
  switch t {
  case .round: return .round
  case .square: return .square
  default: return .butt
  }
}

func mapLineJoin(_ t: RNLineJoinType?) -> CGLineJoin {
  switch t {
  case .round: return .round
  case .bevel: return .bevel
  default: return .miter
  }
}

extension RNPolyline {
  func updatePolyline(_ next: RNPolyline, _ pl: GMSPolyline) {
    let path = GMSMutablePath()
    next.coordinates.forEach {
      path.add(
        CLLocationCoordinate2D(
          latitude: $0.latitude,
          longitude: $0.longitude
        )
      )
    }
    pl.path = path
    if let w = next.width { pl.strokeWidth = CGFloat(w) }
    if let cap = next.lineCap {
      pl.spans = nil
      /// gms.lineCap = mapLineCap(cap)
    }
    if let join = next.lineJoin {
      /// gms.strokeJoin = mapLineJoin(join)
    }
    if let c = next.color?.toUIColor() {
      pl.strokeColor = c
    }
    pl.zIndex = Int32(next.zIndex)
  }

  func polylineEquals(_ b: RNPolyline) -> Bool {
    guard zIndex == b.zIndex,
          (width ?? 0) == (b.width ?? 0),
          lineCap == b.lineCap,
          lineJoin == b.lineJoin,
          color == b.color,
          coordinates.count == b.coordinates.count
    else { return false }
    for i in 0..<coordinates.count {
      if coordinates[i].latitude != b.coordinates[i].latitude
        || coordinates[i].longitude != b.coordinates[i].longitude {
        return false
      }
    }
    return true
  }
}

File: android/src/main/java/com/rngooglemapsplus/MapPolyline.kt
package com.rngooglemapsplus

import com.facebook.react.uimanager.PixelUtil.dpToPx
import com.google.android.gms.maps.model.ButtCap
import com.google.android.gms.maps.model.Cap
import com.google.android.gms.maps.model.JointType
import com.google.android.gms.maps.model.PolylineOptions
import com.google.android.gms.maps.model.RoundCap
import com.google.android.gms.maps.model.SquareCap

class MapPolylineOptions {
  fun buildPolylineOptions(pl: RNPolyline): PolylineOptions =
    PolylineOptions().apply {
      pl.coordinates.forEach { pt ->
        add(
          com.google.android.gms.maps.model
            .LatLng(pt.latitude, pt.longitude),
        )
      }
      pl.width?.let { width(it.dpToPx()) }
      pl.lineCap?.let { startCap(mapLineCap(it)) }
      pl.lineCap?.let { endCap(mapLineCap(it)) }
      pl.lineJoin?.let { jointType(mapLineJoin(it)) }
      pl.color?.let { color(it.toColor()) }
      pl.pressable?.let { clickable(it) }
      pl.zIndex?.let { zIndex(it.toFloat()) }
    }

  fun mapLineCap(type: RNLineCapType?): Cap =
    when (type) {
      RNLineCapType.ROUND -> RoundCap()
      RNLineCapType.SQUARE -> SquareCap()
      else -> ButtCap()
    }

  fun mapLineJoin(type: RNLineJoinType?): Int =
    when (type) {
      RNLineJoinType.ROUND -> JointType.ROUND
      RNLineJoinType.BEVEL -> JointType.BEVEL
      RNLineJoinType.MITER -> JointType.DEFAULT
      null -> JointType.DEFAULT
    }
}

fun RNPolyline.polylineEquals(b: RNPolyline): Boolean {
  if (zIndex != b.zIndex) return false
  if (pressable != b.pressable) return false
  if ((width ?: 0.0) != (b.width ?: 0.0)) return false
  if (lineCap != b.lineCap) return false
  if (lineJoin != b.lineJoin) return false
  if (color != b.color) return false
  val ac = coordinates
  val bc = b.coordinates
  if (ac.size != bc.size) return false
  for (i in ac.indices) {
    val p = ac[i]
    val q = bc[i]
    if (p.latitude != q.latitude || p.longitude != q.longitude) return false
  }
  return true
}
